File: app-ios/VelaWallet/VelaWallet/Features/Onboarding/Core/RegistryResolver.swift
//
//  RegistryResolver.swift
//  VelaWallet
//
//  The three layers behind a sign-in read (064): the index, the registry
//  contract on Gnosis, and its backup on Ethereum.
//
//  Every rule — which layer is asked, when the index's answer is PROVED against
//  the contract's content hash, when it is thrown away, whose unit ids a listing
//  speaks — is `vela_core::registry_resolve`. What is left here is what only a
//  shell can do: perform an `eth_call`. `RegistryClient` performs the index GET
//  and drives the walk.
//

import Foundation
import VelaCore

struct RegistryResolver: Sendable {
    /// The RAW `result` hex, or `nil` when that chain did not answer.
    let ethCall: @Sendable (_ chainId: Int, _ to: String, _ data: String) async -> String?

    /// The core's walks. Replaceable only so a test can script them.
    var keyStep: @Sendable (_ publicKeyHex: String, _ answersJson: String) -> String = {
        registryResolveKeyStep(publicKeyHex: $0, answersJson: $1)
    }
    var unitStep: @Sendable (_ unitId: UInt32, _ source: String, _ answersJson: String) -> String = {
        registryResolveUnitStep(unitId: $0, source: $1, answersJson: $2)
    }
}

File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/onboarding/core/RegistryResolver.kt
package app.getvela.wallet.feature.onboarding.core

/**
 * The three layers — what [RegistryClient] needs to run
 * `vela_core::registry_resolve` (067): the index for speed, the chain for
 * truth, Ethereum for survival.
 *
 * WHICH is asked, in what order, and whether an index answer is believed — it
 * is not: its `contentHash` is recomputed and compared with the chain's — are
 * the core's, written once for four shells. Until 067 this shell carried its
 * own copy of the ladder (`RegistryChainReader`), which asked the chain only
 * when the index was SILENT and believed any index that answered.
 *
 * The three functions are seams so a test can run the REAL core over recorded
 * contract bytes, or script a walk.
 */
class RegistryResolver(
    /** The RAW `result` hex, or `null` when that chain did not answer. */
    val ethCall: suspend (chainId: Int, to: String, data: String) -> String?,
    val keyStep: (publicKeyHex: String, answersJson: String) -> String =
        { key, answers -> uniffi.vela_core_uniffi.registryResolveKeyStep(key, answers) },
    val unitStep: (unitId: Long, source: String, answersJson: String) -> String =
        { id, source, answers -> uniffi.vela_core_uniffi.registryResolveUnitStep(id.toUInt(), source, answers) },
)
